lib/hets-agent/subscriber.rb

```diff
@@ -0,0 +1,115 @@
+require 'json'
+require 'rest-client'
+
+module HetsAgent
+  # Delivers queues for messages and decision which queues should be subscribed
+  class Subscriber
+    def initialize(connection = Bunny.new)
+      @connection = connection
+    end
+
+    def call
+      @connection.start
+      subscribe_version_requirement_queue
+    ensure
+      @connection.close
+    end
+
+    # get version from HetsInstance and parse it
+    def hets_version
+      @hets_version ||= HetsAgent::Hets::VersionCaller.new.call
+    end
+
+    private
+
+    # Binds queue to exchange and subscribes to mininmal parsing version queue
+    def subscribe_version_requirement_queue
+      q_version_requirement = version_requirement_queue
+      q_version_requirement.bind('ex_hets_version_requirement')
+      q_version_requirement.
+        subscribe(block: true,
+                  timeout: 0) do |_delivery_info, _properties, requirement|
+        subscribe_worker_queue(requirement)
+      end
+    end
+
+    # Creates an exchange and a queue for minimal parsing version
+    def version_requirement_queue
+      channel = @connection.create_channel
+      channel.exchange('ex_hets_version_requirement',
+                       type: 'x-recent-history',
+                       durable: true,
+                       arguments: {'x-recent-history-length' => 1})
+      channel.queue("q_hets_version_requirement-#{HetsAgent::Application.id}",
+                    auto_delete: true)
+    end
+
+    # Subscribes to worker queue if min version is <= own version
+    def subscribe_worker_queue(requirement)
+      return unless version_requirement_satisfied?(requirement)
+      queue = create_worker_queue(requirement)
+      print_listening(requirement) if print?
+      queue.subscribe(block: false, manual_ack: true,
+                      timeout: 0) do |delivery_info, _properties, body|
+        queue.channel.acknowledge(delivery_info.delivery_tag)
+        call_hets(JSON.parse(body))
+      end
+    end
+
+    def version_requirement_satisfied?(requirement)
+      Gem::Requirement.new(*requirement.split(',')).
+        satisfied_by?(Gem::Version.new(hets_version))
+    end
+
+    # Creates a worker queue depending on the min parsing version
+    def create_worker_queue(requirement)
+      channel = @connection.create_channel
+      channel.prefetch(1)
+      channel.queue(worker_queue_name(requirement), auto_delete: true)
+    end
+
+    def worker_queue_name(requirement)
+      "hets #{requirement}"
+    end
+
+    def call_hets(data)
+      case data['action']
+      when 'analysis'
+        call_hets_analysis(data['arguments'])
+      when 'version'
+        HetsAgent::Hets::VersionCaller.new.call
+      else
+        $stderr.puts %(Unrecognized action: "#{data['action']}") if print?
+        nil
+      end
+    end
+
+    def call_hets_analysis(arguments)
+      symbolized_arguments = {}
+      %i(file_path file_version_id repository_slug revision server_url
```

Comment:
`arguments.map_keys(&:to_sym).select(...)` might be easier to understand

Reply:
Done in f5a5b1e, but without `map_keys` because it is not a method on a plain Ruby Hash.

Reply:
Oh, does `map_keys` come from ActiveSupport too?

Reply:
Maybe, at least this is not part of plain Ruby.